# 수집된 뉴스 기사 및 댓글에 대한 감정 분석
## * Doc2Vec 
* 데이터 
> 2017년 12월 1일부터 2018년 2월 1일까지 63일간 [네이버](http://www.naver.com)와 [다음](http://www.daum.net)의 랭킹뉴스와 뉴스의 댓글을 크롤링함.

In [1]:
import pickle
import html
import multiprocessing
from collections import namedtuple, OrderedDict
import re
import sys
import os
from glob import glob
import warnings

os.environ['KERAS_BACKEND']='tensorflow'

from numba import jit
import numpy as np
from tqdm import tqdm
tqdm.pandas(desc="progress-bar")
import pandas as pd

from gensim.models import doc2vec, KeyedVectors
from gensim.models.doc2vec import TaggedDocument

from konlpy.utils import pprint

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix, classification_report, roc_curve,  accuracy_score
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.preprocessing import scale, MinMaxScaler
from sklearn.manifold import TSNE

import keras.backend.tensorflow_backend as K
from keras.preprocessing import sequence
from keras_tqdm import TQDMCallback, TQDMNotebookCallback
from keras.models import Model, Sequential
from keras.regularizers import l2
from keras.optimizers import SGD
from keras.utils import np_utils
from keras.layers import Input, Flatten, Dense, Embedding, embeddings, merge, Dropout, Activation,  LSTM, Bidirectional, SimpleRNN, GRU
from keras.layers.convolutional import Conv1D, Conv2D
from keras.layers.pooling import MaxPooling1D, GlobalMaxPooling1D
from keras.layers.normalization import BatchNormalization
from keras.layers.core import SpatialDropout1D
from keras.utils import np_utils
from tensorflow.python.client import device_lib
from keras.layers.merge import dot

import xgboost as xgb

import matplotlib.pyplot as plt

import bokeh.plotting as bp
from bokeh.models import HoverTool, BoxSelectTool
from bokeh.plotting import figure, show, output_notebook

Using TensorFlow backend.


In [2]:
import Database_Handler as dh
import Basic_Module as bm

In [4]:
from ckonlpy.tag import Twitter
from konlpy.tag import Mecab
ct = Twitter()
mecab = Mecab()

## Stopwords

In [5]:
stopwords = open('./data/stopwordsList.txt',encoding='utf-8').readlines()
stopwords = list(map(lambda x: x.strip(), stopwords))

## Load Data

### Path

In [34]:
if sys.platform =='darwin':
    'daum_news_sentiment_analysis.csv'
    loadModelPath = '/Volumes/disk1/model/'
    classifierPath = '/Volumes/disk1/data/pre_data/classifier/'
    news_senti_outcome = '/Volumes/disk1/outcome_for_News_sentiment_analysis/'
    daumCommentPath = './data/daum_Comments/'
    naverCommentsPath = './data/naver_Comments/'
elif sys.platform =='win32':
    loadModelPath = 'd:/model/'
    classifierPath = 'd:/data/pre_data/classifier/'
    newsPath = './data/pre_data/news_sentiment/'
    news_senti_outcome = './outcome_for_News_sentiment_analysis/'
    daumCommentPath = './data/daum_Comments/'
    naverCommentsPath = './data/naver_Comments/'

### News

In [7]:
os.listdir(news_senti_outcome)

['naver_news_sentiment_analysis.csv', 'daum_news_sentiment_analysis.csv']

In [8]:
# Naver
naverData = pd.read_csv(os.path.join(news_senti_outcome, 'naver_news_sentiment_analysis.csv'), index_col=0, header= 0, encoding = 'utf-8')
naverData['site'] = ['Naver'] * naverData.shape[0]
#naverData.head()

In [9]:
naverData[naverData.number_of_crawled_comment == 0]

category        date press  number_of_comment  \
id                                                                       
5a32e268588c136a189e66d0    IT/과학  2017.12.14  노컷뉴스                  1   
5a44bcf3588c13269cc40b83    생활/문화  2017.12.27  연합뉴스                  0   
5a6668ee588c131c881e0a35    IT/과학  2018.01.19   뉴스1                  0   

                          number_of_crawled_comment  rank  \
id                                                          
5a32e268588c136a189e66d0                          0    28   
5a44bcf3588c13269cc40b83                          0    29   
5a6668ee588c131c881e0a35                          0    21   

                                                      title  \
id                                                            
5a32e268588c136a189e66d0               드론, 하천측량 업무에 본격 활용된다   
5a44bcf3588c13269cc40b83        임실 변천사 한눈에…1730년 작성한 운수지 공개   
5a6668ee588c131c881e0a35  CJ헬로 매각 부인한 CJ오쇼핑 '3개월내 매각못한다'…왜?   

                                                                   mainText  \
id                                                                            
5a32e268588c136a189e66d0  국토부, 드론기반 하상변동조사·하천측량 시범사업 착수  [CBS노컷뉴스 곽영식 기자...   
5a44bcf3588c13269cc40b83  (임실=연합뉴스) 이윤승 기자 = 1730년대 전북 임실군의 변천사를 한눈에 파악할...   
5a6668ee588c131c881e0a35  © News1 안은나 기자   계속 제기되는 매각설에 CJ오쇼핑과 CJ헬로 '곤혹'...   

                                         keywords  \
id                                                  
5a32e268588c136a189e66d0  ['드론', '시범사업', '국토교통부']   
5a44bcf3588c13269cc40b83           ['조선시대', '달리']   
5a6668ee588c131c881e0a35  ['이슈 · LGU+, CJ헬로 인수설']   

                                                         extracted_keywords  \
id                                                                            
5a32e268588c136a189e66d0        {'기존', '활용', '드론', '시범 사업', '국토부', '하천 측량'}   
5a44bcf3588c13269cc40b83  {'중요한', '작성', '참여', '운수', '설명', '확보', '임실군', '...   
5a6668ee588c131c881e0a35              {'CJ 헬로', '공시', '결정', 'CJ오쇼핑', '매각설'}   

                          negative  positive  Decision   site  
id                                                             
5a32e268588c136a189e66d0      28.0      80.0  positive  Naver  
5a44bcf3588c13269cc40b83      38.0      70.0  positive  Naver  
5a6668ee588c131c881e0a35      85.0      23.0  negative  Naver

naver 뉴스를 크롤링한 데이터베이스에서는 댓글수가 0개인 뉴스가 3개 존재함

#### 크롤링된 댓글수가 0개인 것을 찾아보자

In [10]:
# Daum
daumData = pd.read_csv(os.path.join(news_senti_outcome, 'daum_news_sentiment_analysis.csv'), index_col=0, header= 0, encoding = 'utf-8')
daumData['site'] = ['daum'] * daumData.shape[0]
#daumData.head()

#### 크롤링된 댓글수가 0개인 것을 찾아보자

In [11]:
daumData[daumData.number_of_crawled_comment == 0]

category        date  press  number_of_comment  \
id                                                                        
5a3c39e7588c132bc05289fb       뉴스  2017.12.21   노컷뉴스              17384   
5a64385d588c134738c38086       연예  2018.01.07    뉴스엔                  0   
5a6472e8588c131b18d57abe      스포츠  2018.01.11  MK스포츠                  0   

                          number_of_crawled_comment  rank  \
id                                                          
5a3c39e7588c132bc05289fb                          0     7   
5a64385d588c134738c38086                          0    45   
5a6472e8588c131b18d57abe                          0    38   

                                                         title  \
id                                                               
5a3c39e7588c132bc05289fb      [인터뷰] 서민 "'문빠' 해악, 더 두고 볼 수 없었다"   
5a64385d588c134738c38086  '언니네' 송은이 "라이언전, 10년 전 박소현 맞선남으로 출연"   
5a6472e8588c131b18d57abe        MLB 에이전트, 선수 샤워 장면 몰래 촬영해 '충격'   

                                                                   mainText  \
id                                                                            
5a3c39e7588c132bc05289fb  - 낚싯배 사고 & 中 기자폭행 관련 '문빠' 반응, 도를 넘은 듯 - '보수 쪽도...   
5a64385d588c134738c38086  [뉴스엔 박수인 기자] 개그우먼 송은이가 작곡가 라이언전의 과거를 공개했다. 1월 ...   
5a6472e8588c131b18d57abe  [매경닷컴 MK스포츠 김재호 기자] 다수의 메이저리그 선수를 고객으로 보유하고 있는...   

                                                        keywords  \
id                                                                 
5a3c39e7588c132bc05289fb                    ['정관용', '댓글', '낚싯배']   
5a64385d588c134738c38086  ['박소현', '언니네라디오', '김숙', '개그우먼', '게스트']   
5a6472e8588c131b18d57abe                                     NaN   

                                                         extracted_keywords  \
id                                                                            
5a3c39e7588c132bc05289fb  {'대통령', '정관용', '서민 교수', '낚싯배', '댓글', '문빠들', '폭행'}   
5a64385d588c134738c38086  {'언니네', '민망', '방송', '박소현 맞선남', '송은이', '작곡가 라이언전'}   
5a6472e8588c131b18d57abe  {'메이저리그', '에이전시', '회사', '사장', 'CSE', '샤워', '선수...   

                          negative  positive  Decision  site  
id                                                            
5a3c39e7588c132bc05289fb      95.0      13.0  negative  daum  
5a64385d588c134738c38086      65.0      43.0  negative  daum  
5a6472e8588c131b18d57abe      39.0      69.0  positive  daum

daum 뉴스를 크롤링한 데이터베이스에서는 댓글수가 0개인 뉴스가 3개 존재함.
> 뉴스 1건의 댓글수(글쓴이가 삭제한 댓글 포함)가 17,384건이지만, 크롤링된 댓글은 0개............  
>> 2017.12.21 뉴스. 발견은 2018년 3월 14일. 3달이나 지나서 댓글을 수집하면 삭제할 사람은 삭제했을 것. 
>> * 아쉽지만 데이터베이스에 댓글이 존재하는 것에 대해서만 진행하기로 함. 

### 새롭게 정의된 뉴스목록

In [18]:
reNaverData = naverData[naverData.number_of_crawled_comment != 0]
reDaumData = daumData[daumData.number_of_crawled_comment != 0]

In [19]:
reDaumData.head()

category        date  press  number_of_comment  \
id                                                                        
5a2a61bf588c13481c229d1e       뉴스  2017.12.07   세계일보               1093   
5a2a61bf588c13481c229d1f       뉴스  2017.12.07   헬스조선                603   
5a2a61bf588c13481c229d20       뉴스  2017.12.07   연합뉴스               1067   
5a2a61bf588c13481c229d21       뉴스  2017.12.07  헤럴드경제                418   
5a2a61bf588c13481c229d22       뉴스  2017.12.07   연합뉴스                434   

                          number_of_crawled_comment  rank  \
id                                                          
5a2a61bf588c13481c229d1e                        903     1   
5a2a61bf588c13481c229d1f                        386     2   
5a2a61bf588c13481c229d20                        807     3   
5a2a61bf588c13481c229d21                        368     4   
5a2a61bf588c13481c229d22                        367     5   

                                                        title  \
id                                                              
5a2a61bf588c13481c229d1e     "밤이 무섭다"..비아그라 공장 연기에 남성들 부작용 호소   
5a2a61bf588c13481c229d1f  식후 커피·늦은 양치질..점심식사 후 하면 안 좋은 습관 3가지   
5a2a61bf588c13481c229d20  '십년지기 생매장' 진짜 이유는.."'청부 통정' 알려질까 봐"   
5a2a61bf588c13481c229d21                       신영자, 억 소리나는 갑질   
5a2a61bf588c13481c229d22   "배신하지마" 20대女 살인 피의자 유치장서 공범 남친에 쪽지   

                                                                   mainText  \
id                                                                            
5a2a61bf588c13481c229d1e   주민들은 공장에서 배출된 연기가 '남성이 매우 건강해지는 부작용'을 일으킨다며, ...   
5a2a61bf588c13481c229d1f   점심식사를 마친 후 후식으로 커피를 마시는 사람들이 많다. 실제로 직장이 밀집돼 ...   
5a2a61bf588c13481c229d20   (성남=연합뉴스) 최해민 기자 = 십년지기 지인을 산 채로 묻어 살해한 50대 여...   
5a2a61bf588c13481c229d21  신영자, 적용안된 혐의→검찰 상고에서 인정\n신영자, 얼마를 어떻게 받았나  [헤럴...   
5a2a61bf588c13481c229d22  (청주=연합뉴스) 이승민 기자 = 지난 9월 청주의 한 하천에서 20대 여성을 둔기...   

                                         keywords  \
id                                                  
5a2a61bf588c13481c229d1e  ['부작용', '비아그라', '아일랜드']   
5a2a61bf588c13481c229d1f      ['커피', '낮잠', '음식물']   
5a2a61bf588c13481c229d20   ['살인혐의', '철원', '검찰송치']   
5a2a61bf588c13481c229d21   ['신영자', '갑질', '롯데백화점']   
5a2a61bf588c13481c229d22       ['공범', '살인', '과자']   

                                                         extracted_keywords  \
id                                                                            
5a2a61bf588c13481c229d1e    {'지역', '남성들', '건강', '공장', '연기', '부작용', '세보 효과'}   
5a2a61bf588c13481c229d1f  {'입냄새', '건강', '자세', '철분', '낮잠', '점심 식사', '치아',...   
5a2a61bf588c13481c229d20  {'진술', '경찰', '남편', '아들', '성관계', '주변', '지인', '철...   
5a2a61bf588c13481c229d21  {'롯데', '혐의', '매장', '검찰', '유통업체', '징역', '신영자 이사...   
5a2a61bf588c13481c229d22  {'과자', '혐의', '쪽지', '남자친구', '유치장', '경찰', '범행', ...   

                          negative  positive  Decision  site  
id                                                            
5a2a61bf588c13481c229d1e      50.0      58.0  positive  daum  
5a2a61bf588c13481c229d1f      81.0      27.0  negative  daum  
5a2a61bf588c13481c229d20      99.0       9.0  negative  daum  
5a2a61bf588c13481c229d21      77.0      31.0  negative  daum  
5a2a61bf588c13481c229d22     102.0       6.0  negative  daum

In [38]:
def Make_Comments_File(filepath, row):
    import Basic_Module as bm
    import os
    filename = row.name
    absPath = os.path.join(filepath, filename+'.csv')
    if os.path.isfile(absPath):
        pass
    else:
        comments = bm.Read_Comments(row)
        return comments
        #comments.to_csv(absPath, index=None, header=True, encoding='utf-8')
    
    

In [39]:
x = Make_Comments_File(daumCommentPath, x)

(903, 8)


In [40]:
x

_id category  \
0    5a2a61c1588c13481c22a047       뉴스   
1    5a2a61c1588c13481c22a0e9       뉴스   
2    5a2a61c1588c13481c229dde       뉴스   
3    5a2a61c1588c13481c22a0f8       뉴스   
4    5a2a61c1588c13481c22a13a       뉴스   
5    5a2a61c1588c13481c22a04b       뉴스   
6    5a2a61c1588c13481c229e8b       뉴스   
7    5a2a61c1588c13481c229e4d       뉴스   
8    5a2a61c1588c13481c22a11f       뉴스   
9    5a2a61c1588c13481c229ed2       뉴스   
10   5a2a61c1588c13481c229f04       뉴스   
11   5a2a61c1588c13481c22a100       뉴스   
12   5a2a61c1588c13481c22a09a       뉴스   
13   5a2a61c1588c13481c229f06       뉴스   
14   5a2a61c1588c13481c22a09c       뉴스   
15   5a2a61c1588c13481c229f05       뉴스   
16   5a2a61c1588c13481c22a0b8       뉴스   
17   5a2a61c1588c13481c229f14       뉴스   
18   5a2a61c1588c13481c229e4b       뉴스   
19   5a2a61c1588c13481c229dd5       뉴스   
20   5a2a61c1588c13481c22a11a       뉴스   
21   5a2a61c1588c13481c22a0dc       뉴스   
22   5a2a61c1588c13481c22a075       뉴스   
23   5a2a61c1588c13481c229f34       뉴스   
24   5a2a61c1588c13481c22a03d       뉴스   
25   5a2a61c1588c13481c229ecf       뉴스   
26   5a2a61c1588c13481c229eb3       뉴스   
27   5a2a61c1588c13481c229e26       뉴스   
28   5a2a61c1588c13481c22a039       뉴스   
29   5a2a61c1588c13481c229f6c       뉴스   
..                        ...      ...   
873  5a2a61c1588c13481c22a045       뉴스   
874  5a2a61c1588c13481c22a097       뉴스   
875  5a2a61c1588c13481c229e25       뉴스   
876  5a2a61c1588c13481c229e5c       뉴스   
877  5a2a61c1588c13481c229de2       뉴스   
878  5a2a61c1588c13481c229ed8       뉴스   
879  5a2a61c1588c13481c229e6d       뉴스   
880  5a2a61c1588c13481c22a065       뉴스   
881  5a2a61c1588c13481c229ef4       뉴스   
882  5a2a61c1588c13481c229fa6       뉴스   
883  5a2a61c1588c13481c229de7       뉴스   
884  5a2a61c1588c13481c229dec       뉴스   
885  5a2a61c1588c13481c229f6e       뉴스   
886  5a2a61c1588c13481c22a0e8       뉴스   
887  5a2a61c1588c13481c229fea       뉴스   
888  5a2a61c1588c13481c229ea2       뉴스   
889  5a2a61c1588c13481c229e55       뉴스   
890  5a2a61c1588c13481c229e37       뉴스   
891  5a2a61c1588c13481c22a04e       뉴스   
892  5a2a61c1588c13481c22a0ed       뉴스   
893  5a2a61c1588c13481c229dce       뉴스   
894  5a2a61c1588c13481c229de3       뉴스   
895  5a2a61c1588c13481c229f9d       뉴스   
896  5a2a61c1588c13481c229f38       뉴스   
897  5a2a61c1588c13481c229f91       뉴스   
898  5a2a61c1588c13481c229ec4       뉴스   
899  5a2a61c1588c13481c229e22       뉴스   
900  5a2a61c1588c13481c22a0c3       뉴스   
901  5a2a61c1588c13481c22a11d       뉴스   
902  5a2a61c1588c13481c229eda       뉴스   

                                              comments        date rank  site  \
0    #옵션열기는 누굽니까!! 윤석렬검사에 강장제와 꽃을 보내자!! 그리고 무일이... ...  2017.12.07    1  daum   
1                                      '"매우 건강"ㅎㅎㅎㅎㅎㅎㅎ  2017.12.07    1  daum   
2    '남성이 매우 건강해지는 부작용'을 일으킨다며 ㅋㅋ 너무 건강해져도 탈이것네.. 시...  2017.12.07    1  daum   
3                                                    .  2017.12.07    1  daum   
4    1. 미국은 한반도 분단 책임지고 한국의 평화통일에 사심없이 적극협조해야 합니다. ...  2017.12.07    1  daum   
5    20대 후반부터 습관적으로 비아그라 즐기던 친구 30대 후반에 발기부전 오더라.. ...  2017.12.07    1  daum   
6                                          24시간 꼴려있나부네  2017.12.07    1  daum   
7                                     24시간 꼴리면 죽을 맛이지?  2017.12.07    1  daum   
8                                          24시간 풀발기 ㅜㅜ  2017.12.07    1  daum   
9                                          24시간발기상태유지.  2017.12.07    1  daum   
10                                              503 보내  2017.12.07    1  daum   
11                         503... 저기면 영서니 안맥이고 할수 있겠다.  2017.12.07    1  daum   
12                                           503가고 싶겠다  2017.12.07    1  daum   
13                              503번에게  귀가 솔깃한  기사로군 ~  2017.12.07    1  daum   
14                                    503에 추천할수 있는곳...  2017.12.07    1  daum   
15                                      503을 저기에 유배시켜라  2017.12.07    1  daum   
16                                         5

In [47]:
x[~x.comments.str.match('.*[0-9a-zA-Z가-힣ㄱ-하-ㅣ]+')]

_id category comments        date rank  site 공감 비공감
3  5a2a61c1588c13481c22a0f8       뉴스        .  2017.12.07    1  daum  0   0

## TaggedDocument For Doc2Vec

In [13]:
TaggedDocument = namedtuple('TaggedDocument', 'words tags sentiment')

## Doc2Vec Model

In [ ]:
model1 = doc2vec.Doc2Vec.load(loadModelPath+'doc2vec_size-1000_epoch-20_window-10_negative-7_hs-0_dm-1_dm_concat-0_dm_mean-1_by-ct.model')
name1 = '-'.join(re.split('[\(\),\/]',str(model1)))+taggerName
model2 = doc2vec.Doc2Vec.load(loadModelPath+'doc2vec_size-1000_epoch-20_window-5_negative-7_hs-0_dm-1_dm_concat-1_dm_mean-0_by-ct.model')
name2 = '-'.join(re.split('[\(\),\/]',str(model2)))+taggerName
model3 = doc2vec.Doc2Vec.load(loadModelPath+'doc2vec_size-1000_epoch-20_window-None_negative-7_hs-0_dm-0_dm_concat-0_dm_mean-0_by-ct.model')
name3 = '-'.join(re.split('[\(\),\/]',str(model3)))+taggerName